### PR TITLE
Remove pipecode and code from sequence exec (replaces #1619)

### PR DESCRIPTION
**Pull request by rabbah**
Replaces #1619.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 1]
```diff
@@ -38,24 +38,19 @@ import whisk.core.entity.size.SizeOptionString
...
- *
- * exec: { kind  : one of supported language runtimes
- *         code  : code to execute if kind is supported
- *         image : container name when kind is "blackbox",
- *         main  : a fully-qualified class name when kind is "java" }
+ * An example exec looks like this:
+ * { kind  : one of supported language runtimes
+ *   code  : code to execute if kind is supported
+ *   image : container name when kind is "blackbox",
+ *   main  : a fully-qualified class name when kind is "java" }
```
Comment (psuter) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: That's not accurate anymore (it's on me, though). It should say something like: "The name of the entrypoint function, when using a non-default value. For Java, the name of the main class."

File: core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala [hunk 2]
```diff
@@ -559,7 +560,7 @@ class ContainerPool(
             try {
                 info(this, s"making new container because none available")
                 startingCounter.next()
-                makeGeneralContainer(key, containerName, imageName, limits, action.exec.pull)
+                makeGeneralContainer(key, containerName, imageName, limits, action.exec.asInstanceOf[CodeExec[_]].pull)
```
Comment (psuter) [inline core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala]: That seems bold? `makeWhiskContainer` doesn't get called for non-JS actions? If so can we add a `require` at the top? (or at least a comment similar to the other new ones)

File: core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala [hunk 3]
```diff
@@ -111,6 +111,9 @@ class Invoker(
         // if the doc revision is missing, then bypass cache
         WhiskAction.get(entityStore, actionid.id, actionid.rev, fromCache = actionid.rev != DocRevision()) onComplete {
             case Success(action) =>
+                // only Exec instances that are subtypes of CodeExec reach the invoker
+                assume(action.exec.isInstanceOf[CodeExec[_]])
```
Comment (psuter) [inline core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala]: all these type checks are making me think we should type-parametrize `WhiskAction[ExecType]` or something, and have methods defined as `def foo[T <: CodeExec](a: WhiskAction[T])`

File: core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala (hunk 2 above)
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala]: `CodeExec` is the root for all actions other than sequences which includes non-js actions and blackbox.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala]: Note the `assume(action.exec.isInstanceOf[CodeExec[_]])` in Invoker.scala asserts this cast is ok.

File: core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala (hunk 3 above)
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala]: I agree in principle but that's a bigger change to stage separately IMO.

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala (hunk 1 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: eh - I read it and decided to leave it. So it's on both of us.
I'll amend.

**Comment by psuter** [comment]
What's the PG id?

**Comment by rabbah** [comment]
oops - sorry. pg 2/857.

**Comment by csantanapr** [comment]
@rabbah This means that we can remove [pipe.js](https://github.com/openwhisk/openwhisk-catalog/blob/master/packages/system/pipe.js) and the system package from the catalog?

cc @houshengbo @jasonpet

**Comment by rabbah** [comment]
This has been true for a while - pipe.js has not been used for sequences in a long time.

**Comment by psuter** [comment]
Sqashed and merged as 7159ef9bed6bc446c534251820be241c1256a27d

**Pull request closed without merging**